sum.cu

```diff
@@ -150,7 +150,55 @@ __global__ void sumKernel3(int4 *d_in, int *d_out, int n) {
     }
 }
 
-const int NUM_KERNELS = 3;
+inline int __device__ warpReduce3(int sum) {
+    int sum_warp = 0;
+    // https://docs.nvidia.com/cuda/parallel-thread-execution/#parallel-synchronization-and-communication-instructions-redux-sync
+    // Use `redux.sync` to reduce within warp
+    asm volatile("redux.sync.add.s32 %0, %1, 0xff;" : "=r"(sum_warp) : "r"(sum));
+    return sum_warp;
+}
+
+template <int BlockSize, int Batch>
+__global__ void sumKernel4(int4 *d_in, int *d_out, int n) {
+    __shared__ __align__(16) int sdata[1];
+
+    unsigned int tid = threadIdx.x;
+    unsigned int i = blockIdx.x*BlockSize*Batch + tid;
+
+    if (i == 0) *d_out = 0;
+    if (tid == 0) sdata[0] = 0;
+
+    int4 value;
+    if (i < n) {
+        // https://docs.nvidia.com/cuda/parallel-thread-execution/#data-movement-and-conversion-instructions-ld
```

Comment:
Lets use `__ldcg` for brevity.

https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html?highlight=__ldcg#load-functions-using-cache-hints